foxai/explainer/computer_vision/algorithm/lrp_rules.py

```diff
@@ -0,0 +1,77 @@
+"""Based on https://github.com/pytorch/captum/blob/master/captum/attr/_core/lrp.py"""
+
+from abc import ABCMeta, abstractmethod
+from typing import Optional
+
+import torch
+from torch.nn import Module, Parameter
+
+
+class PropagationRule(metaclass=ABCMeta):
+    """
+    Base class for all propagation rule classes, also called Z-Rule.
+    epsilon is used to assure that no zero divison occurs.
+    """
+
+    def __init__(self, epsilon=1e-9) -> None:
+        self.epsilon = epsilon
+        self.relevance_input: Optional[torch.Tensor] = None
+        self.relevance_output: Optional[torch.Tensor] = None
+
+    def create_backward_hook_input(self, input_tensor: torch.Tensor):
+        def _backward_hook_input(grad):
+            relevance = grad * input_tensor
+            self.relevance_input = relevance.data
+            return relevance
+
+        return _backward_hook_input
+
+    def create_backward_hook_output(self, outputs):
+        def _backward_hook_output(grad):
+            sign = torch.sign(outputs)
+            sign[sign == 0] = 1
```

Comment:
Is this faster than running: 
`sign = (output > 0).type(torch.uint8)`

Reply:
It would be a nice shortcut, although it is not the same. Original sign is:
1 if output >= 0, -1 otherwise

Your proposition gives 1 >= 0, 0 otherwise
and still 0 will need to be converted to -1.